services/lambda-label-successful-prs/CIBot.py

```diff
@@ -0,0 +1,199 @@
+                 jenkins_url=os.environ.get("jenkins_url"),
+                 jenkins_username=None,
+                 jenkins_password=None,
+                 apply_secret=True):
+        """
+        Initializes the CI Bot
+        :param repo: GitHub repository that is being referenced
+        :param github_personal_access_token: GitHub authentication token (Personal access token)
+        :param jenkins_url: Jenkins URL
+        :param jenkins_username: Jenkins Username for JenkinsAPI
+        :param jenkins_password: Jenkins Password for JenkinsAPI
+        :param apply_secret: GitHub secret credential (Secret credential that is unique to a GitHub developer)
+        """
+        self.repo = repo
+        self.github_personal_access_token = github_personal_access_token
+        self.jenkins_username = jenkins_username
+        self.jenkins_password = jenkins_password
+        if apply_secret:
+            self._get_secret()
+        self.all_jobs = None
+        self.jenkins_url = jenkins_url
+        self.prs_labeled = 0
+        self.prs_already_labeled = 0
+        self.prs_status_fail = 0
+        self.prs_unknown_job = 0
+        self.prs_no_build = 0
+
+    def _get_secret(self):
+        """
+        This method is to get secret value from Secrets Manager
+        """
+        secret = json.loads(secret_manager.get_secret())
+        self.github_personal_access_token = secret["github_personal_access_token"]
+        self.jenkins_username = secret["jenkins_username"]
+        self.jenkins_password = secret["jenkins_password"]
+
+    def _get_github_object(self):
+        """
+        This method returns github object initialized with Github personal access token
+        """
+        github_obj = Github(self.github_personal_access_token)
+        return github_obj
+
+    def _find_all_open_prs(self, github_obj):
+        """
+        This method identifies all the open PRs in the given repository
+        It returns a Paginated list of Pull Request objects sorted in
+        descending order of created date i.e. latest PR first
+        :param github_obj
+        """
+        repo = github_obj.get_repo(self.repo)
+        pulls = repo.get_pulls(state='open', sort='created', direction='desc')
+        return pulls
+
+    def _get_jenkins_obj(self):
+        """
+        This method returns an object of Jenkins instantiated using username, password
+        """
+        return Jenkins(self.jenkins_url, username=self.jenkins_username, password=self.jenkins_password)
+
+    def _get_checks(self):
+        """
+        This method finds out all the checks that are currently supported as part of CI
+        """
+        # for now hardcoding list of checks
+        # ideally use Jenkins API to query list of checks and parse it (bit complicated)
+        all_checks = ['clang', 'edge', 'centos-cpu', 'centos-gpu', 'windows-cpu', 'windows-gpu', 'miscellaneous', 'unix-cpu', 'unix-gpu', 'website', 'sanity']
```

Comment:
Don't query Jenkins but query the GitHub API. Every PR states which checks are required - even before they are run. That list is always up to date and immediately available.

Reply:
I had considered that option of not using Github API. I looked into the PyGithub API documentation.
For every object of type PullRequest, I coudln't find a method that returns the checks corresponding to that PR.

E.g. https://pygithub.readthedocs.io/en/latest/github_objects/PullRequest.html

Hence I had to rely on Jenkins API

Reply:
https://github.com/PyGithub/PyGithub/issues/1063 is an open issue for integrating PyGithub with Checks API. I had created a similar issue yesterday: https://github.com/PyGithub/PyGithub/issues/1621

Reply:
I could use curl/requests to make REST API calls to GitHub [instead of Jenkins]. Before doing that, I wanted to verify: The reasoning behind not using Jenkins API and instead stick to GitHub API is because 
a. Jenkins API can be a bit restrictive. For e.g. There's a round-about way to get status of the Jenkins check
b. Instead of querying Jenkins server on AWS instance for details, we can query GitHub instead.

Are there other compelling reasons to use GitHub API over Jenkins API? Just asking for my knowledge.

Reply:
Checks API is the wrong API. You need the status API: https://docs.github.com/en/rest/reference/repos#statuses

Yes, Jenkins API can be a bit restrictive - also you won't cover other checks (e.g. Travis). GitHub already has the information, so there's no need to ask the source.

Yes, you can query the GitHub servers instead.

Biggest reason is that the GitHub API is the actual source of truth for our users. It doesn't matter if something in Jenkins is green or not - if that information is not present in GitHub, you can't merge. So if you forgot to add a Jenkins check or aren't querying Travis, they might still not be able to merge. The GitHub API knows which checks are required, which ones are present etc. Jenkins on the other hand does not.

Reply:
Btw, you want a webhook based on the following event: https://developer.github.com/webhooks/event-payloads/#status

Reply:
If we are to use the webhook approach, can we request Apache Infra team to update the Events parameter to additionally send Statuses. 

Our previous request to the Infra team had following events:
```
Events : Issue Comments, Pull Requests, Pushes 
```
Refer: https://issues.apache.org/jira/browse/INFRA-20005

Can you help me with that request @marcoabreu Thanks.

Reply:
Why not simply add a new webhook to keep the things separate?

Happy to assist.

Reply:
Please don't forget to write a proposal on dev@ which describes what you're trying to do, what impact it's gonna have and what you need.

Reply:
Do we really need a separate webhook when the 2 webhooks have everything in common except 1 event?

Previous webhook
```
URL : <>
Content type: application/json
Secret : <>
Events : Issue Comments, Pull Requests, Pushes
Enable SSL verification
```
New webhook
```
URL : <same url>
Content type: application/json
Secret : <>
Events : Issue Comments, Pull Requests, Pushes, Statuses
Enable SSL verification
```

Reply:
Well if one thing breaks, do you want everything else to immediately break as well? It's just separation of concerns to reduce the blast radius and reduce the complexity of the application.

One bot is made for reacting to the comments.

This bot is made for reacting to status events.